### PR TITLE
fix(plugin-meetings): trigger event based on H264 presence

### PR DESCRIPTION
**Pull request by abailey-dev**
Developers need an event to know that meetings won't work due to H264 missing in the browser, as well as an event once H264 support is ready (after the browser downloads the codec).

Will check for the presence of H264 and triggers either '**media:codec:missing**' event or the '**media:codec:loaded**' event to inform developers. If the codec is missing it will periodically check if the codec got downloaded up to a maximum timeout. If the codec gets downloaded before the timeout it triggers the '**media:codec:loaded**' event.

Fixes #SPARK-187493

---

Make sure to have followed the [contributing guidelines](https://github.com/webex/webex-js-sdk/blob/master/CONTRIBUTING.md#submitting-a-pull-request) before submitting.

### CODE REVIEW

File: packages/node_modules/@webex/plugin-meetings/src/meetings/util.js [hunk 1]
```diff
@@ -72,4 +76,84 @@ MeetingsUtil.checkForCorrelationId = (deviceUrl, locus) => {
...
+ * @param {String} codec Media codec.
+ * @returns {Boolean}
+ */
+MeetingsUtil.hasMediaCodec = (codec) => {
+  let hasCodec = false;
+  const video = window.document.createElement('video');
+  const PROBABLY = 'probably';
+
+  if (window.MediaSource) {
+    hasCodec = window.MediaSource.isTypeSupported(codec);
```
Comment (abailey-dev) [inline packages/node_modules/@webex/plugin-meetings/src/meetings/util.js:91]: `window.MediaSource.isTypeSupported(codec)` will return `true` or `false` if the codec is supported.

File: packages/node_modules/@webex/plugin-meetings/src/meetings/util.js [hunk 2]
```diff
@@ -72,4 +76,84 @@ MeetingsUtil.checkForCorrelationId = (deviceUrl, locus) => {
...
+  }
+  else {
+    hasCodec = video.canPlayType(codec) === PROBABLY;
```
Comment (abailey-dev) [inline packages/node_modules/@webex/plugin-meetings/src/meetings/util.js:94]: Modenizer does this exact check to determine media capabilities of a browser. `canPlayType` will return only three values 'probably', 'maybe', or ''. If the return is not 'probably' then it doesn't support the codec. We use this as a fallback, preferring `isTypeSupported` first.

Comment (lak-16) [reply packages/node_modules/@webex/plugin-meetings/src/meetings/util.js:94]: I have a Firefox browser where it fails connecting to call due to H264 issue.  I see same values on browser which supports H264 and on browser which doesn't support H26 too.

window.MediaSource.isTypeSupported('video/mp4; codecs="avc1.42E01E"') - returns true 
video.canPlayType('video/mp4; codecs="avc1.42E01E"') - returns "probably".

Comment (abailey-dev) [reply packages/node_modules/@webex/plugin-meetings/src/meetings/util.js:94]: That means the browser you think doesn't support H264 actually does, it supports H.264 Baseline. There's Baseline, Main and High. Let's see which one it doesn't support.

Run this for me, this is for H.264 High:
`window.MediaSource.isTypeSupported('video/mp4; codecs="avc1.64001E"');`

then run this:
`new RTCPeerConnection().createOffer({offerToReceiveVideo: true}).then( offer => console.log(offer.sdp) )`

File: packages/node_modules/@webex/plugin-meetings/src/meetings/util.js [hunk 3]
```diff
@@ -27,6 +29,7 @@ MeetingsUtil.extractDestination = (destination, type) => {
 
 // ** This is a workaround to fix Firefox and google android not downloading H264 codec on start
 MeetingsUtil.triggerH264Codec = async () => {
```
Comment (adamweeks) [inline packages/node_modules/@webex/plugin-meetings/src/meetings/util.js:31]: Is this function needed any more? Can we replace the call to it with a call to `checkMediaPlayback`?

File: packages/node_modules/@webex/plugin-meetings/src/meetings/util.js [hunk 4]
```diff
@@ -72,4 +75,87 @@ MeetingsUtil.checkForCorrelationId = (deviceUrl, locus) => {
...
+
+/**
+ * Notifies the user whether or not the codec
+ * is present. Will continuously check until
+ * max duration.
+ * @async
+ * @param {Number} firstChecked Timestamp in milliseconds.
+ * @returns {undefined}
+ */
+MeetingsUtil.checkMediaPlayback = async (firstChecked) => {
```
Comment (adamweeks) [inline packages/node_modules/@webex/plugin-meetings/src/meetings/util.js:112]: This function name is a bit ambiguous to me. Let's just call it what it is:
```suggestion
MeetingsUtil.checkH264Support = async (firstChecked) => {
```

**Comment by adamweeks** [comment]
Can you update the README with the new events?

Also, we should remove [this section](https://github.com/webex/webex-js-sdk/blob/530bcd182e9da588096df454c72a38af9786f594/packages/node_modules/@webex/plugin-meetings/internal-README.md#L7) from the internal readme.

File: packages/node_modules/@webex/plugin-meetings/src/meetings/util.js (hunk 4 above)
Comment (abailey-dev) [reply packages/node_modules/@webex/plugin-meetings/src/meetings/util.js:112]: Sounds good to me.

Comment (abailey-dev) [reply packages/node_modules/@webex/plugin-meetings/src/meetings/util.js]: Albeit, there is a function with the same name inside PeerConnectionManager.

File: packages/node_modules/@webex/plugin-meetings/src/meetings/util.js (hunk 3 above)
Comment (abailey-dev) [reply packages/node_modules/@webex/plugin-meetings/src/meetings/util.js]: Probably not, just changed it. Thanks!

**Comment by arun3528** [comment]
@adamweeks @anthonywebb i am little scared that this will be a over burden for the developers to check for codec loaded event,
it would be better to put that as part of ready if its loaded.

If its not then send an event for missing codec and loaded , what do you think ?

**Comment by arun3528** [comment]
PR looks good tho

**Comment by adamweeks** [comment]
@abailey-dev after yesterday's meeting, let's go ahead and make this a real PR and get these changes in. 

Also, please update [this section](https://github.com/webex/webex-js-sdk/blob/master/packages/node_modules/@webex/plugin-meetings/README.md#meetings-events) with the new h264 events.

**Review by adamweeks — APPROVED** [review]
LGTM!

**Pull request merged**